File: examples/python/bufr_set_keys.py
# Copyright 2005-2016 ECMWF.
#
# This software is licensed under the terms of the Apache Licence Version 2.0
# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
#
# In applying this licence, ECMWF does not waive the privileges and immunities
# granted to it by virtue of its status as an intergovernmental organisation
# nor does it submit to any jurisdiction.

#
# Python implementation: bufr_set_keys
#
# Description: how to set different type of keys in BUFR messages.
#
#

import traceback
import sys

from eccodes import *

INPUT = '../../data/bufr/syno_multi.bufr'
OUTPUT = 'bufr_set_keys_test_p.tmp.bufr'
VERBOSE = 1  # verbose error reporting


def example():

    # open bufr file
    fin = open(INPUT)

    # open otput bufr file
    fout = open(OUTPUT, 'w')

    cnt = 0

    # loop for the messages in the file
    while 1:

        # get handle for message
        bufr = codes_bufr_new_from_file(fin)
        if bufr is None:
            break

        print "message: %s" % cnt

        # we need to instruct ecCodes to expand all the descriptors
        # i.e. unpack the data values
        # codes_set(bufr,'unpack',1)

        # This is the place where you may wish to modify the message
        # E.g. we change the centre

        # set centre
        val = 222
        print '  set bufrHeaderCentre to: %d' % val

        key = 'bufrHeaderCentre'
        try:
            print '  %s: %s' % (key, codes_set(bufr, key, val))
        except CodesInternalError as err:
            print 'Error with key="%s" : %s' % (key, err.msg)

        # check bufrHeaderCentre's value
        print '  %s''s new value is: %d' % (key, codes_get(bufr, key))

        # write modified message to output
        codes_write(bufr, fout)

        cnt += 1

        # delete handle
        codes_release(bufr)

    fin.close()
    fout.close()


def main():
    try:
        example()
    except CodesInternalError as err:
        if VERBOSE:
            traceback.print_exc(file=sys.stderr)
        else:
            sys.stderr.write(err.msg + '\n')

        return 1

if __name__ == "__main__":
    sys.exit(main())

File: examples/C/bufr_set_keys.c
/*
 * Copyright 2005-2016 ECMWF.
 *
 * This software is licensed under the terms of the Apache Licence Version 2.0
 * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * In applying this licence, ECMWF does not waive the privileges and immunities granted to it by
 * virtue of its status as an intergovernmental organisation nor does it submit to any jurisdiction.
 */

/*
 * C Implementation: bufr_set_keys
 *
 * Description: how to set different type of keys in BUFR messages.
 *
 */

#include "eccodes.h"

void usage(char* prog) {
    printf("usage: %s infile\n",prog);
    exit(1);
}

int main(int argc,char* argv[])
{
    FILE* in = NULL;
    FILE* out = NULL;

    /* message handle. Required in all the eccodes calls acting on a message.*/
    codes_handle* h=NULL;

    long longVal;
    /*double doubleVal;*/
    int err=0;
    int cnt=0;
    size_t size = 0;
    char* infile = "../../data/bufr/syno_multi.bufr";
    const void *buffer = NULL;

    if (argc != 2) {
        usage(argv[0]);
        return 1;
    }

    in = fopen(infile,"r");
    out = fopen(argv[1],"w");

    if (!in || !out) {
        perror("ERROR: unable to open files");
        if (out) fclose(out);
        if (in) fclose(in);
        return 1;
    }

    /* loop over the messages in the bufr file */
    while ((h = codes_handle_new_from_file(NULL,in,PRODUCT_BUFR,&err)) != NULL || err != CODES_SUCCESS)
    {
        if (h == NULL) {
            printf("Error: unable to create handle for message %d\n",cnt);
            cnt++;
            continue;
        }

        printf("message: %d\n",cnt);

        /* we need to instruct ecCodes to expand the descriptors 
          i.e. unpack the data values */
        /*CODES_CHECK(codes_set_long(h,"unpack",1),0);*/

        /* This is the place where you may wish to modify the message*/
        /*E.g. we change the centre  */

        /* set bufrHeaderCentre */
        longVal=222;
        CODES_CHECK(codes_set_long(h, "bufrHeaderCentre", longVal),0);
        printf("  set bufrHeaderCentre to: %ld\n",longVal);

        /* check bufrHeaderCentre */
        CODES_CHECK(codes_get_long(h,"bufrHeaderCentre",&longVal),0);
        printf("  bufrHeaderCentre's new value is: %ld\n",longVal);


        /* get the modified message in a buffer */
        CODES_CHECK(codes_get_message(h,&buffer,&size),0);

        /* write the buffer to a file */
        if(fwrite(buffer,1,size,out) != size) {
            perror(argv[0]);
            return 1;
        }

        /* delete handle */
        codes_handle_delete(h);

        cnt++;
    }

    fclose(in);
    return 0;
}
